lib/connection.cpp

```diff
@@ -1771,3 +1890,195 @@ QVector<Connection::SupportedRoomVersion> Connection::availableRoomVersions() co
     }
     return result;
 }
+
+#ifdef Quotient_E2EE_ENABLED
+void Connection::Private::loadOutdatedUserDevices()
+{
+    QHash<QString, QStringList> users;
+    for(const auto &user : outdatedUsers) {
+        users[user] += QStringList();
+    }
+    if(currentQueryKeysJob) {
+        currentQueryKeysJob->abandon();
+        currentQueryKeysJob = nullptr;
+    }
+    auto queryKeysJob = q->callApi<QueryKeysJob>(users);
+    currentQueryKeysJob = queryKeysJob;
+    connect(queryKeysJob, &BaseJob::success, q, [this, queryKeysJob](){
+        currentQueryKeysJob = nullptr;
+        const auto data = queryKeysJob->deviceKeys();
+        for(const auto &[user, keys] : asKeyValueRange(data)) {
+            deviceKeys[user].clear();
+            for(const auto &device : keys) {
+                if(device.userId != user) {
+                    qCWarning(E2EE) << "mxId mismatch during device key verification:" << device.userId << user;
+                    continue;
+                }
+                if(!device.algorithms.contains("m.olm.v1.curve25519-aes-sha2") || !device.algorithms.contains("m.megolm.v1.aes-sha2")) {
+                    qCWarning(E2EE) << "Unsupported encryption algorithms found" << device.algorithms;
+                    continue;
+                }
+                if(!verifyIdentitySignature(device, device.deviceId, device.userId)) {
+                    qCWarning(E2EE) << "Failed to verify devicekeys signature. Skipping this device";
+                    continue;
+                }
+                deviceKeys[user][device.deviceId] = device;
+            }
+            outdatedUsers -= user;
+        }
+        saveDevicesList();
+    });
+}
+
+void Connection::encryptionUpdate(Room *room)
+{
+    for(const auto &user : room->users()) {
+        if(!d->trackedUsers.contains(user->id())) {
+            d->trackedUsers += user->id();
+            d->outdatedUsers += user->id();
+            d->encryptionUpdateRequired = true;
+        }
+    }
+}
+
+void Connection::Private::saveDevicesList()
+{
+    if (!cacheState)
+        return;
+
+    QElapsedTimer et;
+    et.start();
+
+    QFile outFile { q->e2eeDataDir() % "/deviceslist.json" };
```

Comment:
Why not to the database? That JSON may be very huge potentially.

Reply:
I'll do that